### PR TITLE
Fix swig vs. AVX2

### PR DESCRIPTION
**Pull request by h-vetinari**
Towards #1711. There's still a test [error](https://github.com/facebookresearch/faiss/issues/1711#issuecomment-787509873) that won't be picked up unless an AVX2-specific CI run is added.

### CODE REVIEW

**Comment by mdouze** [comment]
@beauby  could you take a look?

File: conda/faiss/build-pkg.sh [hunk 1]
```diff
@@ -21,6 +21,7 @@ cmake --build _build_python_${PY_VER}_avx2 -j $CPU_COUNT
 # Build vanilla version (no avx).
 cmake -B _build_python_${PY_VER} \
       -Dfaiss_ROOT=_libfaiss_stage/ \
+      -DFAISS_OPT_LEVEL=avx2 \
```
Comment (beauby) [inline conda/faiss/build-pkg.sh:24]: You might want to switch which build has avx2, to remain consistent with the comments.

File: faiss/python/CMakeLists.txt [hunk 2]
```diff
@@ -61,6 +61,10 @@ swig_add_library(swigfaiss
   SOURCES swigfaiss.swig
 )
 
+if(FAISS_OPT_LEVEL STREQUAL "avx2")
+  set_target_properties(swigfaiss PROPERTIES OUTPUT_NAME "swigfaiss_avx2")
```
Comment (beauby) [inline faiss/python/CMakeLists.txt:65]: This looks like a nice solution. Did you test it?

File: conda/faiss/build-pkg.sh (hunk 1 above)
Comment (h-vetinari) [reply conda/faiss/build-pkg.sh:24]: My bad, I got the order wrong.

File: faiss/python/CMakeLists.txt (hunk 2 above)
Comment (h-vetinari) [reply faiss/python/CMakeLists.txt:65]: I did indeed. :)
That's also how I found out that `OUTPUT_NAME` doesn't affect `swigfaiss.py` (not sure if this should be considered a bug in the swig-interface of CMake, but in any case, it doesn't matter much here, as we're renaming the file already).

You can download an artefact from https://github.com/conda-forge/faiss-split-feedstock/pull/34 and test it yourself.

**Comment by h-vetinari** [comment]
I thinking that there should be an AVX2-specific CI run though (assuming the CI has AVX2-capable agents). There's an error in the test suite under AVX2 when building with this patch, see OP, resp. my comment in #1711.

**Comment by beauby** [comment]
@h-vetinari In an ideal world I agree, but AFAIK there's no way to guarantee a CircleCI instance will have an AVX2-enabled CPU (and manual testing shows that at least some hosts do not have AVX2 CPUs).

By testing here I meant did you manually test that your patch leads to the proper .so being loaded when AVX2 is available?

**Comment by h-vetinari** [comment]
Azure consistently has AVX2-enabled workers on linux + windows for example. The PRs in conda-forge run the test-suite with and without AVX2 enabled, and check that the right `.so`s have been loaded by checking the logging statements from `loader.py`. Though just the fact that the test suite now fails with this patch (where it didn't before) shows that it worked.

It's also possible to inspect the `swigfaiss_avx2.py` in the artefact, which reflects the `.so`-rename. In other words, with this patch, `swigfaiss.py` contains:
```python
# Import the low-level C/C++ module
if __package__ or "." in __name__:
    from . import _swigfaiss_avx2
else:
    import _swigfaiss_avx2
```
whereas before it was picking up the wrong `_swigfaiss.so`

**Comment by beauby** [comment]
@h-vetinari Awesome, the last part is what I had in mind. As it seems that every CircleCI 2.0 host has AVX2 enabled, I'm adding a CI job for AVX2, and I'll add a test to our meta.yaml that basically checks that the correct .so is being loaded (something along the lines of `LD_DEBUG=libs python -c "import faiss" 2>&1 | grep libfaiss_avx2.so`. I'll also look into the failing test on your CI.

**Comment by h-vetinari** [comment]
> something along the lines of `LD_DEBUG=libs python -c "import faiss" 2>&1 | grep libfaiss_avx2.so`

That's not bad a bad idea. To stay cross-platform, I used: 
```
pytest tests --log-file-level=INFO --log-file=log.txt
# ensure that expected logger-messages from loader.py is present;
# avoid final '\n', as well as the first 35 = len('INFO     faiss.loader:loader.py:51 ')
python -c "q = open('log.txt').readlines(); import sys; sys.exit(0 if 'Successfully loaded faiss with AVX2 support.' in [x[35:-1] for x in q] else 1)"
```
Which leverages the success messages from #1683. The downside is that this was just working on `swigfaiss.py`-level, so the problem with the wrong `.so` being loaded from withing that file didn't get flagged.

> I'll also look into the failing test on your CI.

Thanks! :)

**Comment by h-vetinari** [comment]
I cherry-picked https://github.com/facebookresearch/faiss/pull/1717/commits/479abe0bfa514d845c0b5f0efabda801e7b3cadf from #1704, since that PR might take longer, but I'd still like to get rid of that annoying warning.

**Comment by beauby** [comment]
@h-vetinari Would you mind reverting the test changes (and send out a separate PR) in order to keep the history clean? After that we can merge this and #1724.

**Pull request closed without merging**